### PR TITLE
Add `.uname` and `.codepoint` to style characters inline

### PR DESCRIPTION
**Pull request by aphillips**
Adds styles recommended by I18N in
https://www.w3.org/TR/international-specs for the inline naming of characters and code points.

\[I18N-ACTION-1257\]

### CODE REVIEW

File: src/base.css
Comment (r12a) [inline src/base.css]: I'd suggest the following changes to the code.
````

.codepoint bdi {
   font-size: 120%;
   }
.codepoint .uname {
   font-size: 80%;
   margin-inline: 2px;
   letter-spacing: 0.03em;
   text-transform: uppercase;
   }
````

Additional, document-specific CSS styling may be necessary to style the contents of the bdi, usually based on the lang attribute attached to that element. This would apply font-family, and possibly (if the font has unusual metrics) font-size settings.

Comment (aphillips) [reply src/base.css]: Done!

**Review by r12a — APPROVED** [review]

File: src/base.css
Comment (aphillips) [reply src/base.css]: @r12a are you able to merge this?

**Review by fantasai — CHANGES_REQUESTED** [review]
- Please document any additions to the stylesheet in the README.
- It looks like you're trying to use smallcaps here, please actually use smallcaps.

**Comment by aphillips** [comment]
@fantasai noted:

> Please document any additions to the stylesheet in the README.

Thanks for the pointer! I will add that to a revision imminently.

> It looks like you're trying to use smallcaps here, please actually use smallcaps.

Do you mean `font-variant: small-caps`? The problem with that is that the name generally is already uppercase and font-variant small-caps only affects the lowercase letters. (It's certainly the case that the code point notation `U+BEEF` is generally already uppercase). This style has the same _effect_ as small-caps, but on mixed case and uppercase.

Of course, I'm probably missing a feature of CSS I don't know. Happy to change to the right thing.

**Comment by aphillips** [comment]
@fantasai:

> Thanks for the pointer! I will add that to a revision imminently.

Hmm. The file `src/README.html` is static HTML. Am I supposed to edit the boilerplate stuff? Should I just add my stuff? Or worry about the "tragedy of the commons"?

**Comment by fantasai** [comment]
> Do you mean font-variant: small-caps? The problem with that is that the name generally is already uppercase and font-variant small-caps only affects the lowercase letters.

* Option A: `font-variant: all-small-caps` https://www.w3.org/TR/css-fonts-4/#font-variant-caps-prop
* Option B: `text-transform: lowercase; font-variant: small-caps`

**Comment by fantasai** [comment]
> Am I supposed to edit the boilerplate stuff?

Edit whatever needs editing. It's not generated, because this is the template against which the generators are designed. :)

**Comment by aphillips** [comment]
(The readme changes will take me some time. I'll update this PR again when we're ready. The change 244dd86 does ***not*** represent "done")

**Comment by aphillips** [comment]
@fantasai We're contemplating further changes to the styling itself. A [question has some up about using color](https://github.com/w3c/i18n-glossary/pull/34#discussion_r1156101131) to set the character name apart from the text--and specifically *which* color to use. Is there any guidance on how to choose a color to match the general color scheme? Could you maybe suggest a color to use? Or should we avoid it?

**Comment by fantasai** [comment]
@aphillips I don't have a good sense of colors. :( What part of this were you planning to color? Can you post an example of the markup?

**Comment by xfq** [comment]
@fantasai We used teal initially, but since it is the same colour as the link colour in i18n articles, we changed it to brown later.

See https://w3c.github.io/bp-i18n-specdev/#example-example-of-a-character-reference for an example.

**Comment by r12a** [comment]
Actually, i think we're good on this now.  We experimented with various alternatives before settling on brown.

**Comment by aphillips** [comment]
Thanks @fantasai, @xfq, and @r12a. When I get a chance, I will update this PR with our various modifications, including the updated readme.

**Comment by fantasai** [comment]
@r12a Color looks nice afaict :) The rule about not including a space between the character and the U+ code seems weird and counter-intuitive, though. I think it would be better to expect it. The font-size change is also a bit disturbing; could we just make it bold? Maybe something like:

```
.codepoint > bdi {
  font-weight: bold;
}
.codepoint > bdi:hover {
  transform-origin: center;
  scale: 130%;
}
```

Another question I have is whether that outermost `span` should be `code`.

And lastly, shouldn't `<code>` default to `translate=no`?

**Comment by xfq** [comment]
> @r12a Color looks nice afaict :) The rule about not including a space between the character and the U+ code seems weird and counter-intuitive, though. I think it would be better to expect it.

I think so too. However, I think using styles to control is for the ease of customization (because it is difficult to change the width of U+0020).

> The font-size change is also a bit disturbing; could we just make it bold? Maybe something like:
> 
> ```css
> .codepoint > bdi {
>   font-weight: bold;
> }
> .codepoint > bdi:hover {
>   transform-origin: center;
>   scale: 130%;
> }
> ```

I think it's to make the character itself (rather than the code point and Unicode name) stand out. I don't have a strong preference here, let's see what others think. I made a comparison document [here](https://xfq.github.io/testing/uname/), so that people can see the rendered document easily.

> Another question I have is whether that outermost `span` should be `code`.

I don't think the outermost `span` should be `code`, because the character is not code.

> And lastly, shouldn't `<code>` default to `translate=no`?

I think code should default to `translate=no`, especially for the non-comment part of the code, but I didn't check if it's specified in the HTML Standard. If the code is translated, it may have an impact on the semantics of the code. Although variable names can be translated technically most of the time, it is not common to do so in practice. If translation is needed, we can add `translate` for those `code` elements.

**Comment by aphillips** [comment]
I think we would want to avoid `font-weight: bold` for a variety of reasons, not the least of which is that support for many scripts can be limited in terms of different weights/styles of fonts being available. It can also make certain characters look different or more difficult to read (or turn into tofu). This may not affect other people's specifications very much, but it might prove a bother for us. Plain styles/normal weights are generally available in fallback fonts or are the _most_ available when it comes to complex things such as conjuncts in Indic scripts. 

I do like the idea of using `:hover` for the size.

**Comment by aphillips** [comment]
I have updated the readme and the stylesheet to match the changes in specdev. Note that I needed to correct an error in the TOC of the readme.

The section about the character template is slightly different from the section in specdev in that it points to specdev (the one in TR) in a couple of places. I also relaxed the wording about not including a space between `bdi` and `span class=uname`. Ready?

File: src/base.css
Comment (fantasai) [inline src/base.css]: Setting the font family to Arial, monospace seems super weird. What's the justification?

File: src/base.css
Comment (fantasai) [inline src/base.css]: Didn't you want `font-variant: all-small-caps` here?

File: src/base.css
Comment (fantasai) [inline src/base.css]: I don't think the inline font-size change is great. It interferes with understanding the relationship of the glyph to the rest of the text. E.g. for é it makes it look like maybe this is a special lowercase e that is the same height as the capital letters (for some weird Unicode-y reason).

File: src/base.css
Comment (fantasai) [inline src/base.css]: If this is supposed to handle multi-character sequences, should it be .unicode instead of .codepoint?

File: src/base.css
Comment (fantasai) [reply src/base.css]: Btw, if you want to zoom on :hover, I think the code that works is

```
bdi { display: inline-block; }
bdi:hover { transform-origin: center; scale: 150%; }
```

**Comment by fantasai** [comment]
(Filed https://github.com/whatwg/html/issues/9364 fwiw.)

File: src/base.css
Comment (aphillips) [reply src/base.css]: We like how Arial looks and know that it is widely available. It is different from the body font of a spec. When Arial is not available, falling back to a sans serif font is not necessarily going to produce the contrast desired. We dithered about just making the font _be_ monospace.

File: src/base.css
Comment (aphillips) [reply src/base.css]: At one time we did the text transform, but the consensus was that this would mask bugs in the user's input of the Unicode character name. Instead of making the capital letters a style, we simulated small caps on text we presume to be all caps to start with.

File: src/base.css
Comment (aphillips) [reply src/base.css]: Let me try the :hover again. I think I was missing the inline-block part.

File: src/base.css
Comment (aphillips) [reply src/base.css]: Good point. Perhaps `.codepoints`?

**Comment by aphillips** [comment]
We discussed this in the I18N teleconference of 2023-06-01 and decided that using `:hover` and making the inline size consistent with the core text was not the right approach. @r12a noted that we want to emphasize the characters rather than the Unicode names and that the styles proposed have been experimented with fairly extensively.

Note that we're working on automation via ReSpec and Bikeshed so that most spec authors won't use these styles directly. They'll use a much more compact notation.

In the change 0aeea32 I also modified text quoted from a CSS spec to use the character template (the better to demonstrate usage). @fantasai have a look at that and see if these are changes that might go upstream 🙂.

File: src/base.css
Comment (aphillips) [reply src/base.css]: we chose not to use the hover and we think the size is a better way to emphasize the character than fooling around with bold (for the reasons we gave before: bold is not always available)

File: src/base.css
Comment (aphillips) [reply src/base.css]: Changing the class name will break a lot of our markup ☹️ so we decided to leave as-is

File: src/base.css
Comment (fantasai) [reply src/base.css]: It's not quite the same effect: https://www.w3.org/TR/css-fonts-4/#font-variant-caps-prop

> would mask bugs in the user's input of the Unicode character name

Is casing really a likely error? I imagine the most likely error is that it'll be written out in all lowercase or in title caps, both of which wouldn't look terribly wrong either way. And it doesn't seem like it would cause a meaningful loss in information, the uppercasing of these names is basically just a convention, isn't it?

File: src/base.css
Comment (fantasai) [reply src/base.css]: Alternatively you could just let the font family inherit? It's defined to be `sans-serif`, which could be anything sans-serif -- including Arial.

You already have color and capitalization shifts here, I don't think a font family shift is also necessary (particularly given this text is reasonably understandable even if there was no formatting change at all!).

File: src/base.css
Comment (fantasai) [reply src/base.css]: In what way?

File: src/base.css
Comment (aphillips) [reply src/base.css]: Agreed. See below for further changes.

File: src/base.css
Comment (aphillips) [reply src/base.css]: This is super-helpful. I made the change, removing the font-family in favor of `all-small-caps`. When I did that I noticed that the digits in a code point (ASCII decimal digits 0-9) display badly compared to small caps for hex digits A-F, so I also added a `font-variant-numeric` with value `oldstyle-nums`. @r12a: have another look at this and see what you think.

File: src/base.css
Comment (aphillips) [reply src/base.css]: We have many many documents with the class `codepoint` in it. If we rename the class to `codepoints` we'd need to change the markup to match. Of course we could carry a local style of `codepoint` while doing migration......

Ultimately our goal is to automate this so that people don't type in the template but instead generate it from a more compact markup using features in e.g. Bikeshed or ReSpec.

**Comment by aphillips** [comment]
@fantasai @r12a Pinging for another look.

**Comment by fantasai** [comment]
@aphillips Please enable maintainer edit permissions

**Comment by aphillips** [comment]
@fantasai It already is enabled?

image

**Comment by fantasai** [comment]
@aphillips OK, I figured it out. :) I've pushed three things here for you to take a look:

1. Rebased on top of the latest origin repo state. (This was a force-push. You'll need to pull again from your own repo to get the new state locally, I hope it doesn't cause too much problems...)
2. Cleaned up the README to follow the style of the rest of the README, fixed indentation, added subheadings, etc.
3. Tweaked the styling a bit to add a background around the BDI, because the font size change is hard to interpret without something to set it apart from the main text.

Take a look and let me know what you think.

> I don't think the outermost `span` should be `code`, because the character is not code.

Should the inner `span` be `code`?

(I do think the `bdi` should be `samp`, but BDI + SAMP seems a bit excessive...)

> `font-variant-numeric: oldstyle-nums;`

What are you trying to do with this? Capital letters + oldstyle numbers is a very weird, and generally unintended, combination...

**Comment by aphillips** [comment]
@fantasai Thanks. I'll pull again and we'll see if I survive the merge 😀 

`span class=uname` might be `code class=uname`. The Unicode name is kind of code-ish.

I agree that BDI+SAMP would be excessive. An alternative might be `<samp dir=auto lang=??>` on the theory that every character or short character sequence's base direction is probably pretty indicative. @r12a any thoughts?

The `oldstyle-nums` was something I introduced with the current crop of edits. When we use `all-small-caps`, the ASCII digits remain their non-small-cap larger selves (because digits are unicameral). This looks clunky and somewhat ransom-note-like:
image

Whereas "oldstyle" numerals are meant to appear with a similar x-height:
image

Admittedly this is weird styling, because the name is "code" and old style numerals are.... yeah. But functionally the height matching works, making it more attractive?!?!?

**Comment by fantasai** [comment]
@aphillips Worst case, if you don't have anything local, you can delete your local copy of the repo and re-clone from `git@github.com:aphillips/tr-design.git`? :)

Wrt oldstyle nums... that's funny, because I thought all-small-caps was supposed to compensate for that. :) https://www.w3.org/TR/css-fonts-4/#example-06f22977 But maybe not. I should raise this question with Myles Maxfield... I suppose that might be an argument for using text-transform + font-size instead of proper small caps. :/

**Comment by fantasai** [comment]
> on the theory that every character or short character sequence's base direction is probably pretty indicative

Well, whether it is nor not, `<bdi>` defaults its `dir` attribute to `auto` anyway, so it's functionally the same either way!

**Comment by aphillips** [comment]
(returning to `oldstyle-nums`) I removed that from the style. It appears that this non-matching of digits in `all-small-caps` is a gap/bug in Chrome (I didn't test extensively, FF does it correctly), so it's probably better to ask for fixes there than go "olde skool" with the digits.

**Comment by aphillips** [comment]
@r12a Can you review and ensure I didn't miss anything from telecon here? I'd like to merge this per our discussion.

Also, mentioning: w3c/i18n-actions#15

File: src/README.html
Comment (r12a) [inline src/README.html]: The whitespace between the 2 lines needs to be removed.

File: src/README.html
Comment (r12a) [inline src/README.html]: I think this should say:
> For invisible characters (such as control characters) or for whitespace, the character and its surrounding <code translate="no" class="kw">bdi</code> element can be omitted, or the <code translate="no" class="kw">bdi</code> element can be replaced with an image.

File: src/README.html
Comment (r12a) [inline src/README.html]: Needs an example of how to use an image.

File: src/README.html
Comment (r12a) [inline src/README.html]: This needs to be updated to show how to use an image, etc.  I will do that, so that you can copy the text, if you wish.

File: src/README.html
Comment (r12a) [inline src/README.html]: Replace

> Although the elements are shown on separate lines in this section (for clarity), you do not need to ..

with

> Do not ...

File: src/README.html
Comment (r12a) [inline src/README.html]: It's also necessary to declare the font family and size for the language text in the CSS.

File: src/README.html
Comment (r12a) [inline src/README.html]: I suggest to change 'although' to 'but'.

File: src/base.css
Comment (r12a) [inline src/base.css]: i'm not clear why this is expressed as a percentage, when the other two values here are in ems.  Why not 1.25em?

File: src/base.css
Comment (r12a) [inline src/base.css]: Why has the small caps reappeared. I thought we banished that long ago, so that content authors would immediately see if they typed the names in lowercase - making the text better suited for copy/paste.

File: src/base.css
Comment (r12a) [inline src/base.css]: The font size appears to have grown also.

Currently, respec_2022.css has the following:
````
.codepoint bdi {
    line-height: 1em;
    font-size: 1.6em;
    padding-inline: 0.25em;
	}

.uname {
	font-size: 75%;
	letter-spacing:0.03em;
	color: brown;
	}
````

File: src/base.css
Comment (r12a) [inline src/base.css:1384]: Give me a short while to update the editor guidelines, so that you can use the examples here.

**Review by r12a — CHANGES_REQUESTED** [review]

**Comment by r12a** [comment]
Ok, i updated the text at https://www.w3.org/International/i18n-activity/guidelines/editing#codepoints

I also added some CSS to deal with img tags.
````
.codepoint img {
	height: 2em;
	margin-inline: 0.25em;
        vertical-align: bottom;
	}
````

File: src/base.css
Comment (aphillips) [reply src/base.css]: We restored small caps instead of using `font-size` per fantasai. Of course, this turns up the small caps numbers bug in Chrome (hence the now removed oldstyle-nums--we just turn a blind eye to Chrome's icky appearance for now). Going back to font-size would fix both.

File: src/README.html
Comment (aphillips) [reply src/README.html]: I fixed all of these by going to `code` vs. `pre`. I also changed "Unicode_character_name" to "UNICODE CHARACTER NAME ALL IN CAPS" to make clear what we want by modeling good behavior.

File: src/README.html
Comment (aphillips) [reply src/README.html]: okay

**Comment by aphillips** [comment]
Okay, something is wrong here. I pushed a new commit to my fork, but it doesn't show in the PR.

https://github.com/w3c/tr-design/compare/gh-pages...aphillips:tr-design:gh-pages?expand=1

File: src/README.html
Comment (aphillips) [reply src/README.html]: DONE

File: src/README.html
Comment (aphillips) [reply src/README.html]: Not sure what you mean here. Do you want to change this text or is this about something else?

**Comment by aphillips** [comment]
@r12a This should do it.

- Changed `span` to `code` for the uname
- Added back `font-family: Arial, monospace` for uname
- Removed the merge conflict
- Added the image example. 

Have a look and I'll ensure that spec dev mirrors this.

File: src/README.html
Comment (r12a) [reply src/README.html]: I think there needs to be a paragraph to remind people using the markup that they will need to use appropriate styling in their style sheet, where necessary, to correctly render the character in the bdi element in terms of font-family and font-size.

**Review by r12a — CHANGES_REQUESTED** [review]

File: src/README.html
Comment (aphillips) [reply src/README.html]: Please suggest something?

Comment (r12a) [reply src/README.html]: Make that an extension of the above para, rather than a separate para.  Perhaps add as the second sentence:

> Your style sheet may need to indicate the appropriate font-family and font-size for characters in a given language.

Comment (aphillips) [reply src/README.html]: I adapted that thusly:

> In some cases, it might also be necessary for you to adjust the <kbd translate="no">font-family</kbd> and/or <kbd translate="no">font-size</kbd> in your own style sheet for characters or character sequences in a given language.

Comment (r12a) [reply src/README.html]: I think that text sounds good, but i'm not seeing it in the diff.

**Comment by aphillips** [comment]
@r12a pinging for approval 😸

File: src/README.html
Comment (r12a) [inline src/README.html]: I think if might be useful to add "for the content of the bdi element" in there somewhere - just so that people don't style the .codepoint element.

In case it helps, the first para of the editorial guidelines now says:

> Use the following markup to represent a character. For certain languages, it may be necessary to style the bdi element with a font-family and font-size.

**Review by r12a — APPROVED** [review]

**Comment by aphillips** [comment]
@fantasai @r12a I think this is ready to merge? I can't do so myself, either because I don't have an approval from @fantasai or possibly I don't have permission.

**Comment by r12a** [comment]
/me could do this, but thinks he should leave this for @fantasai, since it's not his repo.

**Comment by aphillips** [comment]
@fantasai Can you approve? Thanks!!

**Comment by aphillips** [comment]
@fantasai Pinging to bring to the top of your inbox 😉 

Can we please merge this? I'm blocked on pushing the next version of our best practices for spec authors doc on this.

**Review by r12a — APPROVED** [review]

**Review by fantasai — CHANGES_REQUESTED** [review]
I don't understand what happened to this revision: https://github.com/w3c/tr-design/pull/323/commits/f20c65a3ee55d9c654f5c77fa47aa29aa291f641 It looks like you mostly reverted it, and I don't understand why.

Also, please follow the source-formatting conventions of the document you're editing instead of imposing your own on the sections you edit. (And your indentation is inconsistent anyway.)

**Comment by aphillips** [comment]
@fantasai Thanks for the review.

Over the course of the _many_ revisions in this PR, I18N reverted some of the styling suggestions made above (while keeping some others, such as the use of `code` in place of `span`). Notably, we got rid of small caps, old-style numerals, and text transformations to better ensure that the underlying text is copy-ready (because mixed case will show in the page and thus prompt fixes to the underlying text). This is documented by the template text thusly:

```html
<code class="uname">U+XXXX UNICODE_CHARACTER_NAME_ALL_IN_CAPS</code>
```

The text you find here is verbatim from our best practices doc. I can adjust indentation there to match this document if necessary. I think it is important to keep these in sync. That text is staged as [this PR (preview)](https://deploy-preview-113--bp-i18n-specdev.netlify.app/#char_ref_unicode_template).

**Review by r12a — APPROVED** [review]

**Comment by xfq** [comment]
@deniak? Any comments on this PR?

**Comment by r12a** [comment]
In our last CCS<->I18n meeting, Fantasai repeated her preference for a space between the bdi and .uname tags.  Although i had argued that in properly semantic markup it would be better to omit the space, so that the gap could be styled accurately, i can see that people are likely to get this wrong quite often.  Therefore, i don't object to adding the space as part of the canonical markup.

**Comment by deniak** [comment]
> @deniak? Any comments on this PR?

The PR looks good to me but I'll defer to @fantasai who's much more knowledgeable than me on these topics

**Comment by aphillips** [comment]
@r12a Fair enough, although I suspect most people will generate the markup using the shorthands we're adding to respec and bikeshed. 

I considered making the space optional by rephrasing the sentence about spaces thusly:

>  Spacing and presentation is controlled by styling, so you do not need to include line breaks or a space between the closing <code translate="no" class="kw">bdi</code> and the following <code translate="no" class="kw">code</code> element, although you can if you wish.

But then decided to just make a clean break. I added spaces and removed the spacing from the style sheet. While typing this I realize that I didn't fix the examples, but will add one more push in a few seconds and then we should hopefully be done?

**Comment by fantasai** [comment]
@aphillips What's the benefit of adding images for invisible characters? Is it really worth the cost to a) creating the images b) tracking the images with the specification text c) complicating the stylesheet and markup patterns and d) disrupting the paragraph line spacing?

**Comment by fantasai** [comment]
@aphillips So... I broke something. Can you give me write access to your repo so I can try to fix it? :)

**Comment by aphillips** [comment]
@fantasai I invited you to my fork... have at it 👍 

Regarding images, @r12a has been using them for some time and feels that they provide better usability. The images helps make invisibles clearer, at least in some cases. An argument can be made that the sometimes detract from the page and that it is a maintenance headache (especially with regard to echidna configuration), but it is somewhat optional. We might make it clearer that it is up the discretion of the author to use images if they so desire and to spell out the maintenance tradeoff.

@r12a, @xfq any additional thoughts?

**Review by fantasai — APPROVED** [review]
@aphillips OK, I'm going to suggest we tackle that question separately. I want to land what we've got for the basic markup since this is getting to be a really complicated PR.

**Comment by aphillips** [comment]
@fantasai 

The diff got super icky on the pushes you made, so I'm not sure what you changed, but as far as I can tell this looks good. Okay to merge?

**Comment by aphillips** [comment]
@fantasai Argh. Never mind: you already did merge it. A little blind today.

**Comment by xfq** [comment]
> @aphillips What's the benefit of adding images for invisible characters? Is it really worth the cost to a) creating the images b) tracking the images with the specification text c) complicating the stylesheet and markup patterns and d) disrupting the paragraph line spacing?

It is not only useful for invisible characters, but also useful for combining characters. If the font doesn't support display with a dotted circle (`U+25CC DOTTED CIRCLE`), then an image is needed.

**Comment by xfq** [comment]
@deniak Since we have merged this, I wonder if there is anything we need to do to ship it to /TR specs?

**Comment by deniak** [comment]
> @deniak Since we have merged this, I wonder if there is anything we need to do to ship it to /TR specs?

The TR stylesheet has been updated already. See:
* https://www.w3.org/StyleSheets/TR/2021/base.css
* https://www.w3.org/StyleSheets/TR/2021/README.html

**Comment by xfq** [comment]
Great. Thank you all!

**Comment by r12a** [comment]
The section at https://www.w3.org/StyleSheets/TR/2021/README.html#unicode-codepoints looks out of date to me.

It says:

> For invisible characters (such as control characters) or for whitespace, the character and its surrounding bdi element should be omitted:

 but the latest guidelines should suggest using a graphic, eg.

> For invisible characters (such as control characters) or for whitespace, use an image instead of the character; or you may also omit the character and its surrounding bdi element.
> `<span class="codepoint" translate="no"><span class="uname">U+XXXX UNICODE_CHARACTER_NAME_ALL_IN_CAPS</span></span>`

I haven't checked other bits, but it suggests to me that we probably should to make sure that is the latest copy, throughout.

**Comment by xfq** [comment]
That's because @fantasai reverted that part, but it did make tr-design inconsistent with the recommendation of the i18n specdev document, and it's something we should try to avoid.

**Pull request merged**
